### PR TITLE
Implement stream::cycle(..)

### PR DESCRIPTION
**Pull request by felipesere**
This is a rough initial step that is able to take a vec and cycle it indefinitly.

I saw there were comments on #34 about the API and using `IntoStream`.
Would I have to impl IntoStream for a `Vec<T>` myself to be able to have the example be liek below?

```
use stream::IntoStream;

let vals = vec![1,2,3];
let s = vals.into_stream().cycle();

assert_eq!(s.next().await, Some(1));
assert_eq!(s.next().await, Some(2));
assert_eq!(s.next().await, Some(3));
assert_eq!(s.next().await, Some(1));
assert_eq!(s.next().await, Some(2));
```

I'm happy to get some pointers and see this through :) 

Ref: #129

### CODE REVIEW

File: src/stream/cycle.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+///
+/// assert_eq!(s.next().await, Some(1));
+/// assert_eq!(s.next().await, Some(2));
+/// assert_eq!(s.next().await, Some(3));
+/// assert_eq!(s.next().await, Some(1));
+/// assert_eq!(s.next().await, Some(2));
+/// #
+/// # })
+/// ```
+pub fn cycle<T: Copy>(source: Vec<T>) -> impl Stream<Item = T> {
```
Comment (k-nasa) [inline src/stream/cycle.rs]: I thought Clone was better here, but what about it?

Comment (k-nasa) [inline src/stream/cycle.rs]: Shouldn't it be a stream instead of a vec?

Comment (felipesere) [reply src/stream/cycle.rs]: Oh I thought I had taken Copy from somewhere else.
Good shout though.
I have changed the bound to `Clone` now and added calls to it where the compiler complained 😸

**Comment by felipesere** [comment]
@yoshuawuyts The bounds are adjusted and we take as Stream as argument.

What API changes do you have in mind? Should I attach it to stream itself so once can write

```
let s = stream::once(7).cycle();
```

File: src/stream/cycle.rs [hunk 2]
```diff
@@ -40,21 +71,21 @@ impl<T: Copy> Stream for Cycle<T> {
...
+/// assert_eq!(s.next().await, Some(7));
+/// assert_eq!(s.next().await, Some(7));
+/// assert_eq!(s.next().await, Some(7));
+/// assert_eq!(s.next().await, Some(7));
 /// #
 /// # })
 /// ```
-pub fn cycle<T: Copy>(source: Vec<T>) -> impl Stream<Item = T> {
-    let len = source.len();
+pub fn cycle<S: Stream<Item = T>, T: Clone>(source: S) -> impl Stream<Item = S::Item> {
```
Comment (yoshuawuyts) [inline src/stream/cycle.rs]: This should probably be a method on `Stream`, and not a free function. Similar to [`Iterator::cycle`](https://doc.rust-lang.org/core/iter/trait.Iterator.html#method.cycle)

**Comment by yoshuawuyts** [comment]
@felipesere yes, that's exactly what I was hoping you'd say! -- this should indeed be a method on `Stream`. Very excited for progress on this; we're getting real close to having all methods on `Stream` implemented!

**Comment by felipesere** [comment]
@yoshuawuyts I've moved it, but I think clippy is unhappy on CI and not in a mood to run? 😃

File: src/stream/cycle.rs (hunk 2 above)
Comment (felipesere) [reply src/stream/cycle.rs]: Done 🎉

File: src/stream/stream/cycle.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    }
+}
+
+#[derive(Eq, PartialEq)]
+enum CycleState {
+    FromStream,
+    FromBuffer,
+}
+
+impl<T: Clone, S: Stream<Item = T>> Cycle<S, T> {
```
Comment (k-nasa) [inline src/stream/stream/cycle.rs]: I think this code is better.  I don't feel the need to derive a generic type T.
Do you think?

```rust
impl<S> Cycle<S, S::Item>
where
    S: Stream,
    S::Item: Clone,
{
```

Comment (felipesere) [reply src/stream/stream/cycle.rs]: Is it also possible to use the same trick for the struct definition?

File: src/stream/stream/mod.rs [hunk 4]
```diff
@@ -369,6 +371,38 @@ extension_trait! {
             Chain::new(self, other)
         }
 
+        #[doc = r#"
+            Creats a stream that yields the provided values infinitely and in order.
```
Comment (yoshuawuyts) [inline src/stream/stream/mod.rs]:
```suggestion
            Creates a stream that yields the provided values infinitely and in order.
```

File: src/stream/stream/cycle.rs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            state: CycleState::FromStream,
+        }
+    }
+}
+
+impl<S, T> Stream for Cycle<S, T>
+where
+    S: Stream<Item = T>,
+    T: Clone,
+{
```
Comment (k-nasa) [inline src/stream/stream/cycle.rs]: I don't feel the need to derive a generic type T.
It ’s the same as before.

**Comment by felipesere** [comment]
@k-nasa @yoshuawuyts anything else to add? 😅

**Comment by felipesere** [comment]
@yoshuawuyts anything I should still do on this? :)

**Pull request merged**